lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -767,6 +761,92 @@ mkTipSyncClient tr np localTxSubmissionQ onTipUpdate onPParamsUpdate onInterpret
         NodeToClientV_3
 
 
+-- Reward Account Balances
+
+-- | Monitors values for keys, and allows clients to @query@ them.
+--
+-- Designed to be used for observing reward balances, where we want to query the
+-- balances of /all/ the wallets' accounts on tip change, and allow wallet
+-- workers to @query@ the value.
+--
+-- One could imagine replacing @query@ with a push-based approach.
+data Observer m key value = Observer
+    { startObserving :: key -> m ()
+    , stopObserving :: key -> m ()
+    , query :: key -> m (Maybe value)
+    }
+
+newRewardBalanceFetcher
+    :: forall sc. Tracer IO (NetworkLayerLog sc)
+    -> W.GenesisParameters
+    -- ^ Used to convert tips for logging
+    -> TQueue IO (LocalStateQueryCmd (CardanoBlock sc) IO)
+    -> IO ( Observer IO W.ChimericAccount W.Coin
+          , Tip (CardanoBlock sc) -> IO ()
+            -- Call on tip-change to refresh
+          )
+newRewardBalanceFetcher tr gp queryRewardQ = do
+    cacheVar <- atomically $ newTVar Map.empty
+    toBeObservedVar <- atomically $ newTVar Set.empty
+    return (observer cacheVar toBeObservedVar, refresh cacheVar toBeObservedVar)
+  where
+    observer
+        :: TVar IO (Map W.ChimericAccount W.Coin)
+        -> TVar IO (Set W.ChimericAccount)
+        -> Observer IO W.ChimericAccount W.Coin
+    observer cacheVar toBeObservedVar =
+        Observer
+            { startObserving = \k ->
+                atomically $ do
+                    modifyTVar' toBeObservedVar (Set.insert k)
+            , stopObserving = \k ->
+                atomically $ do
+                    modifyTVar' toBeObservedVar (Set.delete k)
+                    modifyTVar' cacheVar (Map.delete k)
```

Comment:
Integration tests pass, meaning this should work.

But if this `Observer` abstraction is something we want to keep, it would be nice to add more extensive unit tests. `newRewardBalanceFetcher` would probably have to be made more abstract.

I also guess there is similarities between this and the `WorkerRegistry`, that perhaps could be unified in the future.